lib/worker/output.go

```diff
@@ -0,0 +1,219 @@
+type ProcessOutputEntry struct {
+	Content    []byte
+	ReceivedAt time.Time
+}
+
+// Listener represents an output listener that the output handler will
+// send updates to.
+type Listener struct {
+	outputChan chan ProcessOutputEntry
+	errorChan  chan error
+}
+
+// NewOutputHandler returns a new OutputHandler struct from the provided readers.
+func NewOutputHandler(rc ...io.Reader) (*OutputHandler, error) {
+	if len(rc) == 0 {
+		return nil, fmt.Errorf("must provide at least one io.Reader")
+	}
+	o := &OutputHandler{
+		readers:  rc,
+		mu:       &sync.Mutex{},
+		done:     make(chan struct{}),
+		messages: make(chan []byte),
+		errors:   make(chan error),
+	}
+	go o.handleOutput()
+	go o.handleBroadcast()
+	go o.handleErrors()
+	return o, nil
+}
+
+// Stream returns a channel to receive output up to the
+// present time and stream real time. The channel closes when
+// there the process the output is coming from is finished.
+// Any errors encountered during output reading are sent through
+// the errors channel.
+func (o *OutputHandler) Stream() (chan ProcessOutputEntry, chan error) {
+	output := make(chan ProcessOutputEntry)
+	errChan := make(chan error)
+
+	go func() {
+		if o.bufferLen() > 0 {
+			output <- ProcessOutputEntry{Content: o.combinedBuffer}
+		}
+		select {
+		case <-o.done:
+			close(output)
+		default:
+			o.addListener(&Listener{
+				outputChan: output,
+				errorChan:  errChan,
+			})
+		}
+	}()
+
+	return output, errChan
+}
+
+func (o *OutputHandler) addListener(newListener *Listener) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	o.listeners = append(o.listeners, newListener)
+}
+
+func (o *OutputHandler) updateCombinedBuffer(b []byte) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	o.combinedBuffer = append(o.combinedBuffer, b...)
+}
+
+func (o *OutputHandler) bufferLen() int {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return len(o.combinedBuffer)
+}
+
+func (o *OutputHandler) iterListeners(routine func(*Listener)) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	for _, l := range o.listeners {
+		routine(l)
+	}
+}
+
+// handleBroadcast forwads new messages to the listeners and
+// closes listener channels when done signal is received to
+// signal end of stream.
+func (o *OutputHandler) handleBroadcast() {
+L:
+	for {
+		select {
+		case msg := <-o.messages:
+			o.iterListeners(func(l *Listener) {
+				l.outputChan <- ProcessOutputEntry{Content: msg}
+			})
+		case <-o.done:
+			o.iterListeners(func(l *Listener) {
+				close(l.outputChan)
+			})
+			break L
+		}
+	}
+}
+
+// handleErrors forwards errors to the listeners
+func (o *OutputHandler) handleErrors() {
+L:
+	for {
+		select {
+		case err := <-o.errors:
+			o.iterListeners(func(l *Listener) {
+				l.errorChan <- err
+			})
+			break L
+		}
+	}
+}
+
+// handleOutput manages the goroutines that consume
+// the reader to send to buffer and listeners. It
+// closes the done channel done.
+func (o *OutputHandler) handleOutput() error {
+	wg := sync.WaitGroup{}
+
+	readersLen := len(o.readers)
```

Comment:
The `readers` slice is not added to once the `Process` is underway, is it? It's just an abstraction over the process' `stdout` and `stderr`, yeah?

Reply:
Yes that is correct. I've added more to the struct field comment to clarify this.